arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -450,6 +450,13 @@ static int arm_mpu_init(struct device *arg)
 
 	_arm_mpu_config();
 
+	/* Sanity check for number of regions in Cortex-M0+, M3, and M4. */
+#if defined(CONFIG_CPU_CORTEX_M0PLUS) || \
```

Comment:
Not sure the reason for this check, are we verifying that the HW reports a sane number of regions?

Reply:
@andrewboie yes, that's correct.
But let me explain this, more, cause this is an essential part of this PR.
I have modified the _get_num_regions(), so it returns a hard-coded value (8) for CM0+, CM3,CM4. That's because the number of MPU regions in these MCUs is 8. So, it does _not_ query the HW in run-time! The reason for this modification is to accelerate the function, by avoiding run-time register checks.

The __ASSERT here is to check "once and for all in a boot lifetime" that the MCU _does_ actually have 8 regions. This is added, signal an error in case of accidental selection of CPU_HAS_MPU in a Cortex-M0+/3/4.
If anyone makes this error, the __ASSERT will hit, cause the TYPE register will indicate 0 regions, and booting will be aborted. 

The overhead of this solution is minimal; just a few instructions. Furthermore, this code is pre-kernel initialization and is guaranteed to execute early at device boot-up. So there is no way that an erroneously configured image (i.e. with MPU enabled while not supported by HW) will continue to execute.

Reply:
works for me